### PR TITLE
feat: implement option to overwrite cassandra config

### LINKED ISSUE #3455: [cassandra] cassandra.yaml override configmap or cluster crd option
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
There are a lot of options in cassandra.yaml config file. According to the code, we load the file from container, replace some options and put it back.
Code reference here: 
https://github.com/rook/rook/blob/master/pkg/operator/cassandra/sidecar/config.go#L316

But if we want to change another settings, we should build a custom docker image with already changed config file and specify it in cluster crd. Or use another work around method.

**What is use case behind this feature:**

I want to do something like rook-config-override config map, but for cassandra.
After cassandracluster CRD was created, the operator started to write rook-cassandra-config-override config map with settings:
```
"rpc_address" = "0.0.0.0"
"endpoint_snitch" = "GossipingPropertyFileSnitch"
```
Users can fill another override settings in the map. 
Config map will be mounted to the pod and all override settings wil be applied to original cassandra.yaml. 

Configuration for cassandra will become more flexible. 

P.s. 
I want to use same scheme to add rook-cassandra-jmx-config in the PR.
https://github.com/rook/rook/pull/3228

P.s. 
Another way is just store cassandra settings in cluster crd, (like in [strimzi kafka operator](https://strimzi.io/), for example), but I want to do it similar to rook operator ceph module. And I do not want cassandra cluster CRD becomes incredibly big.

**Environment**:
https://github.com/rook/rook/blob/master/pkg/operator/cassandra/sidecar/config.go#L316

### PR DESCRIPTION
**Pull request by moolen**
**Description of your changes:**
This allows a user to modify the `cassandra.yml` using a `ConfigMap`. As of now not all properties can be overridden. This lays the foundation to support encrypted server-to-server and client-to-server communication.

most relevant use-cases:
* support usage with istio
* support for crypto (client-server, server-server, encryption at rest in the future)
* fine-tuning of cassandra parameters

This implementation was inspired by the [nginx-ingress configmap](https://github.com/kubernetes/ingress-nginx/blob/master/internal/ingress/controller/template/configmap.go). 

**Which issue is resolved by this Pull Request:**
Resolves #3455 

~**TODO**~
- [ ] get feedback :wave: 
- [x] ~implement PoC~
- [x] ~create integration tests to ensure templating works as expected :unicorn:~
- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/operator/cassandra/sidecar/config/cassandra.go [hunk 1]
```diff
@@ -0,0 +1,319 @@
+package config
+
+// Configuration represents the content of cassandra.yaml file
+// see http://wiki.apache.org/cassandra/StorageConfiguration for full explanations
+type Configuration struct {
```
[the remark this answers is not in the document]
Comment (moolen) [reply pkg/operator/cassandra/sidecar/config/cassandra.go:6]: good point, I added It. I also did a quick diff of cassandra v2 and v3 configs. Most notably fields were added to the config scheme.

**Comment by galexrt** [comment]
@moolen I'll take another look at the PR tomorrow (Monday).

File: tests/framework/installer/cassandra_manifests.go [hunk 2]
```diff
@@ -258,6 +273,7 @@ spec:
     racks:
       - name: "us-east-1a"
         members: %[2]d
+        configMapName: "cassandra-config-map"
```
Comment (galexrt) [inline tests/framework/installer/cassandra_manifests.go:276]: Please add this line commented out with a comment to the `cluster.yaml` example manifest in the `cluster/` dir and the example in the documentation https://rook.io/docs/rook/v1.3/cassandra-cluster-crd.html#sample.
In addition to that the option should be added to the list on the https://rook.io/docs/rook/v1.3/cassandra-cluster-crd.html#settings-explanation

File: Documentation/cassandra.md [hunk 3]
```diff
@@ -92,6 +92,84 @@ cluster = Cluster(['rook-cassandra-client.rook-cassandra.svc.cluster.local'])
...
+## Configuring Cassandra
+
+In order to overwrite the `cassandra.yaml` configuration values you can specify a configmap using `RackSpec.configMapName`.
+You can add key-value pairs to the data section of the ConfigMap.
+
+### Configuration Options
+
+| Name     | Type    | Default |
+|----------|-------------|---------|
+| `cluster_name` | string | the name of your cluster as specified in the cluster crd |
```
Comment (galexrt) [inline Documentation/cassandra.md:104]: Please add backticks to the "type" column as well.
Additionally please format the markdown table as seen in https://github.com/rook/rook/blob/master/README.md#L80-L88 (e.g., Visual Code has a formatter for Markdown).

File: pkg/operator/cassandra/sidecar/config/cassandra.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	PermissionsValidityInMs    int    `json:"permissions_validity_in_ms"`
+	CredentialsValidityInMs    int    `json:"credentials_validity_in_ms"`
+	Partitioner                string `json:"partitioner"`
+	CdcEnabled                 bool   `json:"cdc_enabled"`
+	DiskFailurePolicy          string `json:"disk_failure_policy"`
+	CommitFailurePolicy        string `json:"commit_failure_policy"`
+	// TODO
+	//PreparedStatementsCacheSizeMb       interface{} `json:"prepared_statements_cache_size_mb"`
+	//ThriftPreparedStatementsCacheSizeMb interface{} `json:"thrift_prepared_statements_cache_size_mb"`
+	//KeyCacheSizeInMb   interface{} `json:"key_cache_size_in_mb"`
```
Comment (galexrt) [inline pkg/operator/cassandra/sidecar/config/cassandra.go:29]: Are these and the other commented out options going to be implemented or removed from the code after review?

**Comment by yanniszark** [comment]
@moolen thanks a lot for your PR! This is awesome work :smile:
I'll try to take a look at the PR this weekend. @galexrt thanks a lot for doing a review pass!

File: pkg/operator/cassandra/sidecar/config/cassandra.go (hunk 4 above)
Comment (moolen) [reply pkg/operator/cassandra/sidecar/config/cassandra.go:29]: The idea was to leave this as a TODO in the code - if someone needs a particular config field one can use the search to find it in the code. I don't know yet how to handle the types for these configuration values (docs say: `Default value is empty to make it "auto"`).

I can remove it or leave a comment with more details if you want

Comment (galexrt) [reply pkg/operator/cassandra/sidecar/config/cassandra.go:29]: I wonder if `0` would be "empty" for Cassandra in that case, then we could have them in there as integers or can it also be a string (e.g., `128M` or similar).

If `0` doesn't mean they are "empty", then let's change the comment to exactly that information, that they are not included to the need of being empty, on the other hand Go's json tags should also allow for `omitempty` which might do the trick here. Could you check that if it is possible?

Comment (moolen) [reply pkg/operator/cassandra/sidecar/config/cassandra.go]: I did research on that: In some cases`0` means disable cache key size:

```yaml
# Maximum size of the key cache in memory.
#
# Each key cache hit saves 1 seek and each row cache hit saves 2 seeks at the
# minimum, sometimes more. The key cache is fairly tiny for the amount of
# time it saves, so it's worthwhile to use it at large numbers.
# The row cache saves even more time, but must contain the entire row,
# so it is extremely space-intensive. It's best to only use the
# row cache if you have hot rows or static rows.
#
# NOTE: if you reduce the size, you may not get you hottest keys loaded on startup.
#
# Default value is empty to make it "auto" (min(5% of Heap (in MB), 100MB)). Set to 0 to disable key cache.
key_cache_size_in_mb:
```
source: taken from cassandra:3.11 image `/etc/cassandra/cassandra.yaml`

`omitempty` does the trick. cassandra does not complain, thank you!

**Comment by moolen** [comment]
BTW: the failing tests relate to #6085 (not a cassandra test failing)

**Comment by moolen** [comment]
ping @yanniszark

**Comment by yanniszark** [comment]
@moolen thanks a lot for the PR.
Sorry for the long wait! I was on PTO and now I'm very short on time.
I will try to review ASAP.
BTW, if you are interested in becoming a maintainer for the Cassandra Operator (if it makes sense, since you're using it), I would love to help you achieve that, as I don't have much time available.

**Comment by moolen** [comment]
Hey @yanniszark, thank you for your trust! Unfortunately i can't commit my time reliably into a maintainer role. Tho i would love to add e2e tests and other functionality to bring the cassandra operator forward.

**Pull request closed without merging**